src/cascadia/TerminalSettingsModel/BaseVisualStudioGenerator.cpp

```diff
@@ -11,10 +11,17 @@ void BaseVisualStudioGenerator::GenerateProfiles(std::vector<winrt::com_ptr<impl
 {
     // There's no point in enumerating valid Visual Studio instances more than once,
     // so cache them for use by both Visual Studio profile generators.
-    static const auto instances = VsSetupConfiguration::QueryInstances();
+    static auto instances = VsSetupConfiguration::QueryInstances();
 
-    for (auto const& instance : instances)
+    // Sort instances based on version and install date.
+    std::sort(instances.begin(), instances.end(), BaseVisualStudioGenerator::Compare);

```

Comment:
Could you move this into `VsSetupConfiguration::QueryInstances`?
That way this function would remain thread safe.

Additionally if you reverse the sort order you can iterate through it with your regular for loop.

Instead of defining an explicit function for `BaseVisualStudioGenerator::Compare` you can write a lambda:
```
std::sort(instances.begin(), instances.end(), [](const auto& a, const auto& b) {
    const auto aVersion = a.GetComparableVersion();
    const auto bVersion = b.GetComparableVersion();

    if (aVersion == bVersion)
    {
        return a.GetComparableInstallDate() > b.GetComparableInstallDate();
    }

    return aVersion > bVersion;
});
```

Reply:
I prototyped with a lambda, but this function is going to likely grow. Personally not a fan of monolithic lambdas but if you prefer that I will.

Reply:
Please keep it as a separate function if you prefer that. 🙂 But I'd like to ask you to move it into `VsSetupConfiguration::QueryInstances`, as that would keep the function thread-safe. I believe reversing the sort order a good idea as well.

Reply:
By the way: Do we need to use stable sort here? If so we should switch to `std::stable_sort`.

Reply:
It's practically impossible when comparing version + install date that you have equivalent instances. The VS setup engine can only install 1 instance at a time and blocks for others until it's finished. Not only did we do that to mitigate issues with machine state changes based on what we detected before execution of our install plan began (impossible to prevent), but some packages we install - like Windows Installer packages (though very few) - will block concurrent installs anyway. We do have retry loops for that, but it makes installs take longer.

As such, and given the relatively low number of instances expected on the vast majority of devices (from telemetry, 1 dwarfs 2 or more instances, and it quickly falls off after 2 and especially 3 instances) I don't think `std::stable_sort` is necessary.

I will move the sort inside `QueryInstances`, though. It's a good idea.